File: src/main/java/com/github/simiacryptus/aicoder/psi/PsiUtil.java
package com.github.simiacryptus.aicoder.psi;

import com.github.simiacryptus.aicoder.util.StringTools;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Stream;

public class PsiUtil {

    public static PsiElement getLargestIntersectingComment(@NotNull PsiElement element, int selectionStart, int selectionEnd) {
        return getLargestIntersecting(element, selectionStart, selectionEnd, "PsiCommentImpl", "PsiDocCommentImpl");
    }

    /**
     * This method is used to get the largest element that intersects with the given selection range.
     *
     * @param element        The element to search within.
     * @param selectionStart The start of the selection range.
     * @param selectionEnd   The end of the selection range.
     * @param types          The types of elements to search for.
     * @return The largest element that intersects with the given selection range.
     */
    public static PsiElement getLargestIntersecting(@NotNull PsiElement element, int selectionStart, int selectionEnd, CharSequence... types) {
        final AtomicReference<PsiElement> largest = new AtomicReference<>(null);
        final AtomicReference<PsiElementVisitor> visitor = new AtomicReference<>();
        visitor.set(new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                if (null == element) return;
                TextRange textRange = element.getTextRange();
                boolean within = (textRange.getStartOffset() <= selectionStart && textRange.getEndOffset() + 1 >= selectionStart && textRange.getStartOffset() <= selectionEnd && textRange.getEndOffset() + 1 >= selectionEnd);
                CharSequence simpleName = element.getClass().getSimpleName();
                if (Arrays.asList(expand(types)).contains(simpleName)) {
                    if (within) {
                        largest.updateAndGet(s -> (s == null ? 0 : s.getText().length()) > element.getText().length() ? s : element);
                    }
                }
                super.visitElement(element);
                element.acceptChildren(visitor.get());
            }
        });
        element.accept(visitor.get());
        return largest.get();
    }
    public static List<PsiElement> getAll(@NotNull PsiElement element, CharSequence... types) {
        final List<PsiElement> elements = new ArrayList<>();
        final AtomicReference<PsiElementVisitor> visitor = new AtomicReference<>();
        visitor.set(new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                if (null == element) return;
                if (Arrays.asList(expand(types)).contains(element.getClass().getSimpleName())) {
                    elements.add(element);
                } else {
                    element.acceptChildren(visitor.get());
                }
                super.visitElement(element);
            }
        });
        element.accept(visitor.get());
        return elements;
    }

    /**
     * This method is used to get the smallest intersecting entity of a given PsiElement.
     *
     * @param element        The PsiElement to search for the smallest intersecting entity.
     * @param selectionStart The start of the selection range.
     * @param selectionEnd   The end of the selection range.
     * @return The smallest intersecting entity of the given PsiElement.
     */
    public static PsiElement getSmallestIntersecting(@NotNull PsiElement element, int selectionStart, int selectionEnd) {
        return getSmallestIntersecting(element, selectionStart, selectionEnd, "PsiMethodImpl", "PsiFieldImpl", "PsiClassImpl");
    }

    /**
     * This method is used to get the smallest intersecting entity from a given PsiElement.
     *
     * @param element        The PsiElement from which the smallest intersecting entity is to be retrieved.
     * @param selectionStart The start of the selection.
     * @param selectionEnd   The end of the selection.
     * @param types          The types of the elements to be retrieved.
     * @return The smallest intersecting entity from the given PsiElement.
     */
    public static PsiElement getSmallestIntersecting(@NotNull PsiElement element, int selectionStart, int selectionEnd, CharSequence... types) {
        final AtomicReference<PsiElement> largest = new AtomicReference<>(null);
        final AtomicReference<PsiElementVisitor> visitor = new AtomicReference<>();
        visitor.set(new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                if (null == element) return;
                TextRange textRange = element.getTextRange();
                boolean within = (textRange.getStartOffset() <= selectionStart && textRange.getEndOffset() + 1 >= selectionStart && textRange.getStartOffset() <= selectionEnd && textRange.getEndOffset() + 1 >= selectionEnd);
                CharSequence simpleName = element.getClass().getSimpleName();
                if (Arrays.asList(expand(types)).contains(simpleName)) {
                    if (within) {
                        largest.updateAndGet(s -> (s == null ? Integer.MAX_VALUE : s.getText().length()) < element.getText().length() ? s : element);
                    }
                }
                //System.out.printf("%s : %s%n", simpleName, element.getText());
                super.visitElement(element);
                element.acceptChildren(visitor.get());
            }
        });
        element.accept(visitor.get());
        return largest.get();
    }

    private static CharSequence[] expand(CharSequence[] types) {
        return Arrays.stream(types).flatMap(x-> Stream.of(x, StringTools.stripSuffix(x, "Impl"))).distinct().toArray(CharSequence[]::new);
    }

    public static PsiElement getFirstBlock(@NotNull PsiElement element, CharSequence blockType) {
        PsiElement[] children = element.getChildren();
        if(null == children || 0 == children.length) return null;
        PsiElement first = children[0];
        if(first.getClass().getSimpleName().equals(blockType)) return first;
        return null;
    }

    public static PsiElement getLargestBlock(@NotNull PsiElement element, CharSequence blockType) {
        AtomicReference<PsiElement> largest = new AtomicReference<>(null);
        AtomicReference<PsiElementVisitor> visitor = new AtomicReference<>();
        visitor.set(new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                CharSequence simpleName = element.getClass().getSimpleName();
                if (simpleName.equals(blockType)) {
                    largest.updateAndGet(s -> s != null && s.getText().length() > element.getText().length() ? s : element);
                    super.visitElement(element);
                } else {
                    super.visitElement(element);
                }
                element.acceptChildren(visitor.get());
            }
        });
        element.accept(visitor.get());
        PsiElement psiElement = largest.get();
        return psiElement;
    }

    /**
     * This method returns a {@link HashSet} of {@link String}s containing the simple names of all the {@link PsiElement}s
     * contained within the given {@link PsiElement}.
     *
     * @param element The {@link PsiElement} whose children's simple names are to be retrieved.
     * @return A {@link HashSet} of {@link String}s containing the simple names of all the {@link PsiElement}s contained
     * within the given {@link PsiElement}.
     */
    public static @NotNull HashSet<CharSequence> getAllElementNames(@NotNull PsiElement element) {
        HashSet<CharSequence> set = new HashSet<>();
        AtomicReference<PsiElementVisitor> visitor = new AtomicReference<>();
        visitor.set(new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                CharSequence simpleName = element.getClass().getSimpleName();
                set.add(simpleName);
                element.acceptChildren(visitor.get());
            }
        });
        element.accept(visitor.get());
        return set;
    }

    public static @NotNull String printTree(@NotNull PsiElement element) {
        StringBuilder builder = new StringBuilder();
        printTree(element, builder, 0);
        return builder.toString();
    }

    private static void printTree(@NotNull PsiElement element, @NotNull StringBuilder builder, int level) {
        for (int i = 0; i < level; i++) {
            builder.append("  ");
        }
        builder.append(element.getClass().getSimpleName() + "    " + element.getText().replaceAll("\n", "\\\\n"));
        builder.append("\n");
        for (PsiElement child : element.getChildren()) {
            printTree(child, builder, level + 1);
        }
    }

    public static PsiElement getLargestContainedEntity(PsiElement element, int selectionStart, int selectionEnd) {
        if(null == element) return element;
        TextRange textRange = element.getTextRange();
        if(textRange.getStartOffset() >= selectionStart && textRange.getEndOffset() <= selectionEnd) return element;
        PsiElement largestContainedChild = null;
        for (PsiElement child : element.getChildren()) {
            PsiElement entity = getLargestContainedEntity(child, selectionStart, selectionEnd);
            if(null != entity) {
                if (largestContainedChild == null || largestContainedChild.getTextRange().getLength() < entity.getTextRange().getLength()) {
                    largestContainedChild = entity;
                }
            }
        }
        return largestContainedChild;
    }

}

File: src/main/kotlin/com/simiacryptus/aicoder/util/psi/PsiUtil.kt
package com.simiacryptus.aicoder.util.psi

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.simiacryptus.util.StringUtil
import java.util.*
import java.util.concurrent.atomic.AtomicReference
import java.util.stream.Collectors
import java.util.stream.Stream

object PsiUtil {
  val ELEMENTS_CODE = arrayOf<CharSequence>(
    "Method", "Field", "Class", "Function", "CssBlock", "FunctionDefinition",
    "Property", "Interface", "Enum", "Constructor", "Parameter", "Variable"
  )
  val ELEMENTS_COMMENTS = arrayOf<CharSequence>(
    "Comment", "DocComment", "LineComment", "BlockComment", "JavadocComment"
  )

  // Common block types used in multiple places
  private val BLOCK_TYPES = arrayOf(
    "CodeBlock", "BlockExpr", "Block", "BlockExpression", "StatementList", "BlockFields",
    "ClassBody", "MethodBody", "FunctionBody", "TryBlock", "CatchBlock", "FinallyBlock"
  )

  /**
   * Gets the name of an element (class, method, field etc)
   */
  fun getName(element: PsiElement): String? {
    if (!matchesType(element, *ELEMENTS_CODE)) return null
    val declaration = getDeclaration(element)
    // Extract name from declaration based on element type
    return when {
      matchesType(element, "Class", "Interface", "Enum") ->
        declaration.substringAfter("class ")
          .substringAfter("interface ")
          .substringAfter("enum ")
          .substringBefore("<")
          .substringBefore(" ")
          .trim()

      matchesType(element, "Method", "Function") ->
        declaration.substringAfter(" ")
          .substringBefore("(")
          .trim()

      matchesType(element, "Field", "Variable") ->
        declaration.substringAfterLast(" ")
          .substringBefore("=")
          .trim()

      else -> null
    }
  }

  fun getAll(element: PsiElement, vararg types: CharSequence): List<PsiElement> {
    val elements: MutableList<PsiElement> = ArrayList()
    val visitor = AtomicReference<PsiElementVisitor>()
    visitor.set(object : PsiElementVisitor() {
      override fun visitElement(element: PsiElement) {
        if (matchesType(element, *types)) {
          elements.add(element)
        }
        element.acceptChildren(visitor.get())
        super.visitElement(element)
      }
    })
    element.accept(visitor.get())
    return elements
  }

  fun getSmallestIntersecting(
    element: PsiElement, selectionStart: Int, selectionEnd: Int, vararg types: CharSequence
  ): PsiElement? {
    val smallest = AtomicReference<PsiElement?>(null)
    val visitor = AtomicReference<PsiElementVisitor>()
    visitor.set(object : PsiElementVisitor() {
      override fun visitElement(element: PsiElement) {
        val textRange = element.textRange
        if (intersects(TextRange(selectionStart, selectionEnd), textRange)) {
          if (matchesType(element, *types)) {
            smallest.updateAndGet { s: PsiElement? ->
              if ((s?.text?.length ?: Int.MAX_VALUE) < element.text.length) s else element
            }
          }
        }
        super.visitElement(element)
        element.acceptChildren(visitor.get())
      }
    })
    element.accept(visitor.get())
    return smallest.get()
  }

  private fun within(textRange: TextRange, vararg offset: Int): Boolean = offset.any { it in textRange.startOffset..textRange.endOffset }
  private fun intersects(a: TextRange, b: TextRange): Boolean {
    return within(a, b.startOffset, b.endOffset) || within(b, a.startOffset, a.endOffset)
  }

  fun matchesType(element: PsiElement, vararg types: CharSequence): Boolean {
    return matchesType(element.javaClass.simpleName, types)
  }

  fun matchesType(simpleName: CharSequence, types: Array<out CharSequence>): Boolean {
    var simpleName1 = simpleName
    simpleName1 = StringUtil.stripSuffix(simpleName1, "Impl")
    simpleName1 = StringUtil.stripPrefix(simpleName1, "Psi")
    val str = simpleName1.toString()
    return Stream.of(*types).map { s: CharSequence? ->
      StringUtil.stripSuffix(
        s!!, "Impl"
      )
    }.map { s: String? ->
      StringUtil.stripPrefix(
        s!!, "Psi"
      )
    }.anyMatch { t: CharSequence -> str.endsWith(t.toString()) }
  }

  fun getFirstBlock(element: PsiElement, vararg blockType: CharSequence): PsiElement? {
    val children = element.children
    if (children.isEmpty()) return null
    val first = children[0]
    return if (matchesType(first, *blockType)) first else null
  }

  fun getLargestBlock(element: PsiElement, vararg blockType: CharSequence): PsiElement? {
    val largest = AtomicReference<PsiElement?>(null)
    val visitor = AtomicReference<PsiElementVisitor>()
    visitor.set(object : PsiElementVisitor() {
      override fun visitElement(element: PsiElement) {
        if (matchesType(element, *blockType)) {
          largest.updateAndGet { s: PsiElement? -> if (s != null && s.text.length > element.text.length) s else element }
          super.visitElement(element)
        } else {
          super.visitElement(element)
        }
        element.acceptChildren(visitor.get())
      }
    })
    element.accept(visitor.get())
    return largest.get()
  }

  fun printTree(element: PsiElement): String {
    val builder = StringBuilder()
    printTree(element, builder, 0)
    return builder.toString()
  }

  private fun printTree(element: PsiElement, builder: StringBuilder, level: Int) {
    builder.append("  ".repeat(0.coerceAtLeast(level)))
    val elementClass: Class<out PsiElement> = element.javaClass
    val simpleName = getName(elementClass)
    builder.append(simpleName).append("    ").append(element.text.replace("\n".toRegex(), "\\\\n"))
    builder.append("\n")
    for (child in element.children) {
      printTree(child, builder, level + 1)
    }
  }

  private fun getName(elementClass: Class<*>): String {
    var elementClassVar = elementClass
    val stringBuilder = StringBuilder()
    val interfaces = getInterfaces(elementClassVar)
    while (elementClassVar != Any::class.java) {
      if (stringBuilder.isNotEmpty()) stringBuilder.append("/")
      stringBuilder.append(elementClassVar.simpleName)
      elementClassVar = elementClassVar.superclass
    }
    stringBuilder.append("[ ")
    stringBuilder.append(interfaces.stream().sorted().collect(Collectors.joining(",")))
    stringBuilder.append("]")
    return stringBuilder.toString()
  }

  fun getInterfaces(elementClass: Class<*>): Set<String> {
    val strings = Arrays.stream(elementClass.interfaces).map { obj: Class<*> -> obj.simpleName }.collect(
      Collectors.toCollection { HashSet() })
    if (elementClass.superclass != Any::class.java) strings.addAll(getInterfaces(elementClass.superclass))
    return strings
  }

  fun getLargestContainedEntity(element: PsiElement?, selectionStart: Int, selectionEnd: Int): PsiElement? {
    if (null == element) return null
    val textRange = element.textRange
    if (textRange.startOffset >= selectionStart && textRange.endOffset <= selectionEnd) return element
    var largestContainedChild: PsiElement? = null
    for (child in element.children) {
      val entity = getLargestContainedEntity(child, selectionStart, selectionEnd)
      if (null != entity) {
        if (largestContainedChild == null || largestContainedChild.textRange.length < entity.textRange.length) {
          largestContainedChild = entity
        }
      }
    }
    return largestContainedChild
  }

  fun getLargestContainedEntity(e: AnActionEvent): PsiElement? {
    val caret = e.getData(CommonDataKeys.CARET) ?: return null
    var psiFile: PsiElement? = e.getData(CommonDataKeys.PSI_FILE) ?: return null
    val selectionStart = caret.selectionStart
    val selectionEnd = caret.selectionEnd
    val largestContainedEntity = getLargestContainedEntity(psiFile, selectionStart, selectionEnd)
    if (largestContainedEntity != null) psiFile = largestContainedEntity
    return psiFile
  }

  fun getSmallestContainingEntity(element: PsiElement?, selectionStart: Int, selectionEnd: Int, minSize: Int = 0): PsiElement? {
    if (null == element) {
      return null
    }
    for (child in element.children) {
      val entity = getSmallestContainingEntity(child, selectionStart, selectionEnd, minSize)
      if (null != entity) {
        return entity
      }
    }
    val textRange = element.textRange
    if (textRange.startOffset <= selectionStart) {
      if (textRange.endOffset >= selectionEnd) {
        if (element.text.length >= minSize) {
          return element
        }
      }
    }
    return null
  }

  fun getCodeElement(
    psiFile: PsiElement?, selectionStart: Int, selectionEnd: Int
  ) = getSmallestIntersecting(psiFile!!, selectionStart.toInt(), selectionEnd.toInt(), *ELEMENTS_CODE)

  fun getDeclaration(element: PsiElement): String {
    var declaration: CharSequence = element.text
    declaration = StringUtil.stripPrefix(declaration.toString().trim { it <= ' ' }, getDocComment(element).trim { it <= ' ' })
    declaration = StringUtil.stripSuffix(declaration.toString().trim { it <= ' ' }, getCode(element).trim { it <= ' ' })
    return declaration.toString().trim { it <= ' ' }
  }

  fun getCode(element: PsiElement): String {
    val codeBlock = getLargestBlock(element, *BLOCK_TYPES)
    var code = ""
    if (null != codeBlock) {
      code = codeBlock.text
    }
    return code
  }

  fun getDocComment(element: PsiElement): String {
    var docComment = getLargestBlock(element, *ELEMENTS_COMMENTS)
    if (null == docComment) docComment = getFirstBlock(element, *ELEMENTS_COMMENTS)
    return docComment?.text?.trim() ?: ""
  }

}